jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java

```diff
@@ -294,6 +294,22 @@ public void setForcedPath(String forcedPath)
         _forcedPath = forcedPath;
     }
 
+    private void copyClassServlet(ServletHolder holder) throws ServletException
+    {
+        if (holder.getClassName() != null)
+            this.setClassName(holder.getClassName());
+        // TODO adding this after setting the classname (or instead of) fixes the exceptions.
+        /*
+        Class<? extends Servlet> servletClass = holder.getHeldClass();
+        if (servletClass != null)
+            this.setHeldClass(servletClass);
+        */
+        if (holder.getServlet() != null)
+            this.setServlet(holder.getServlet());
+        if (holder.getServletInstance() != null)
+            this.setInstance(holder.getServletInstance());
```

Comment:
@janbartel going to need some help on this one!

The fields that I think we need to consider copying are:
```
ServletHolder:
    Servlet _servlet
BaseHolder:
    Source _source
    Class<? extends T> _class
    String _className;
    T _instance;
```

The `_instance` and `_servlet` are the the same value, although `_serlvet` can be null until `initServlet()`, so it kind of records the init state as something separate to the started state.  If `getServlet()` is called once the handler is running, then `initServlet()` will be called.

The current code copies `_className` and `_servlet`, but not `_source` or `_class`... or any of the config fields.
So this method is a little conflicted?   Is it named to say it copies just the class, but it will copy an actual instance if it is there.  But not the config within the holder... so there is a risk that a single instance will be initialised twice with different configurations.

If different holders have the same `_servlet`, then it will be destroyed multiple times.

So we have to think what we are really trying to achieve here?  Do we want to copy just the class or the actual instance?    I think if we are doing this for real JSPs then we want to copy to instance as we don't want multiple instances of the `JspServlet`.    In which case we have to be careful about: exactly when the copy is done; what we do with conflicting config; how we stop multiple inits and destroys;

Reply:
In the case where the jsp_file has _not_ been precompiled into a servlet, the existing code copies all of the init params of the JspServlet over. 

The only code change we need to made is to copy the _class_ instead of it's _name_.

Reply:
resolved after conversation with @janbartel